File: src/api/src/routes/capital.ts
import { Router, Request, Response } from "express";
import pool from "../db/index.js";

const router = Router();

interface CreateCapitalSupplyRequest {
  userId: number;
  token: string;
  amount: number;
  walletAddress: string;
  txHash?: string;
}

interface GetCapitalSuppliesQuery {
  userId?: string;
  token?: string;
  status?: string;
}

interface GenerateDepositAddressRequest {
  userId: number;
  token: string;
}

// Helper function to generate a fake Ethereum address
function generateFakeEthAddress(): string {
  const chars = "0123456789abcdef";
  let address = "0x";
  for (let i = 0; i < 40; i++) {
    address += chars[Math.floor(Math.random() * chars.length)];
  }
  return address;
}

router.get("/", async (req: Request<object, object, object, GetCapitalSuppliesQuery>, res: Response) => {
  try {
    const { userId, token, status } = req.query;

    let query = `
      SELECT
        id,
        user_id,
        token,
        amount,
        wallet_address,
        tx_hash,
        status,
        created_at,
        updated_at
      FROM capital_supplies
      WHERE 1=1
    `;
    const params: (string | number)[] = [];
    let paramCount = 1;

    if (userId) {
      query += ` AND user_id = $${paramCount}`;
      params.push(Number(userId));
      paramCount++;
    }

    if (token) {
      query += ` AND token = $${paramCount}`;
      params.push(token);
      paramCount++;
    }

    if (status) {
      query += ` AND status = $${paramCount}`;
      params.push(status);
      paramCount++;
    }

    query += ` ORDER BY created_at DESC`;

    const result = await pool.query(query, params);

    const supplies = result.rows.map(row => ({
      id: row.id,
      userId: row.user_id,
      token: row.token,
      amount: parseFloat(row.amount),
      walletAddress: row.wallet_address,
      txHash: row.tx_hash,
      status: row.status,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
    }));

    res.json(supplies);
  } catch (error) {
    console.error("Error fetching capital supplies:", error);
    res.status(500).json({ error: "Failed to fetch capital supplies" });
  }
});

router.post("/", async (req: Request<object, object, CreateCapitalSupplyRequest>, res: Response) => {
  try {
    const { userId, token, amount, walletAddress, txHash } = req.body;

    if (!userId || !token || !amount || !walletAddress) {
      res.status(400).json({
        error: "userId, token, amount, and walletAddress are required"
      });
      return;
    }

    // Validate token
    const validTokens = ["AAUD", "USDC", "USDT"];
    if (!validTokens.includes(token)) {
      res.status(400).json({
        error: `Invalid token. Must be one of: ${validTokens.join(", ")}`
      });
      return;
    }

    const result = await pool.query(
      `INSERT INTO capital_supplies (user_id, token, amount, wallet_address, tx_hash, status)
       VALUES ($1, $2, $3, $4, $5, 'pending')
       RETURNING id, user_id, token, amount, wallet_address, tx_hash, status, created_at`,
      [userId, token, amount, walletAddress, txHash || null]
    );

    const supply = result.rows[0];

    console.log(`Created capital supply ${supply.id} for user ${userId}: ${amount} ${token}`);

    res.status(201).json({
      id: supply.id,
      userId: supply.user_id,
      token: supply.token,
      amount: parseFloat(supply.amount),
      walletAddress: supply.wallet_address,
      txHash: supply.tx_hash,
      status: supply.status,
      createdAt: supply.created_at,
    });
  } catch (error) {
    console.error("Error creating capital supply:", error);
    res.status(500).json({ error: "Failed to create capital supply" });
  }
});

router.post("/deposit-address", async (req: Request<object, object, GenerateDepositAddressRequest>, res: Response) => {
  try {
    const { userId, token } = req.body;

    if (!userId || !token) {
      res.status(400).json({
        error: "userId and token are required"
      });
      return;
    }

    // Validate token
    const validTokens = ["AAUD", "USDC", "USDT"];
    if (!validTokens.includes(token)) {
      res.status(400).json({
        error: `Invalid token. Must be one of: ${validTokens.join(", ")}`
      });
      return;
    }

    // Check if user already has an active deposit address for this token
    const existingAddress = await pool.query(
      `SELECT id, address, created_at FROM deposit_addresses
       WHERE user_id = $1 AND token = $2 AND status = 'active' AND swept = FALSE
       ORDER BY created_at DESC LIMIT 1`,
      [userId, token]
    );

    if (existingAddress.rows.length > 0) {
      // Return existing address
      const address = existingAddress.rows[0];
      res.json({
        id: address.id,
        address: address.address,
        token,
        createdAt: address.created_at,
        isNew: false,
      });
      return;
    }

    // Generate a new fake Ethereum address
    const depositAddress = generateFakeEthAddress();

    const result = await pool.query(
      `INSERT INTO deposit_addresses (user_id, token, address, status, swept)
       VALUES ($1, $2, $3, 'active', FALSE)
       RETURNING id, user_id, token, address, status, swept, created_at`,
      [userId, token, depositAddress]
    );

    const newAddress = result.rows[0];

    console.log(`Generated deposit address ${depositAddress} for user ${userId}: ${token}`);

    res.status(201).json({
      id: newAddress.id,
      userId: newAddress.user_id,
      token: newAddress.token,
      address: newAddress.address,
      status: newAddress.status,
      swept: newAddress.swept,
      createdAt: newAddress.created_at,
      isNew: true,
    });
  } catch (error) {
    console.error("Error generating deposit address:", error);
    res.status(500).json({ error: "Failed to generate deposit address" });
  }
});

router.get("/deposit-addresses", async (req: Request<object, object, object, { userId?: string }>, res: Response) => {
  try {
    const { userId } = req.query;

    if (!userId) {
      res.status(400).json({ error: "userId is required" });
      return;
    }

    const result = await pool.query(
      `SELECT id, user_id, token, address, status, swept, created_at, updated_at
       FROM deposit_addresses
       WHERE user_id = $1
       ORDER BY created_at DESC`,
      [Number(userId)]
    );

    const addresses = result.rows.map(row => ({
      id: row.id,
      userId: row.user_id,
      token: row.token,
      address: row.address,
      status: row.status,
      swept: row.swept,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
    }));

    res.json(addresses);
  } catch (error) {
    console.error("Error fetching deposit addresses:", error);
    res.status(500).json({ error: "Failed to fetch deposit addresses" });
  }
});

export default router;

File: src/api_go/handlers/capital.go
package handlers

import (
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"fmt"
	"log"
	"net/http"
	"strconv"

	"paperhands/api/config"
	"paperhands/api/models"

	"github.com/gin-gonic/gin"
)

var validTokens = []string{"AAUD", "USDC", "USDT"}

type CreateCapitalSupplyRequest struct {
	UserID        int     `json:"userId" binding:"required"`
	Token         string  `json:"token" binding:"required"`
	Amount        float64 `json:"amount" binding:"required"`
	WalletAddress string  `json:"walletAddress" binding:"required"`
	TxHash        string  `json:"txHash"`
}

type GenerateDepositAddressRequest struct {
	UserID int    `json:"userId" binding:"required"`
	Token  string `json:"token" binding:"required"`
}

func isValidToken(token string) bool {
	for _, t := range validTokens {
		if t == token {
			return true
		}
	}
	return false
}

func generateEthAddress() string {
	bytes := make([]byte, 20)
	rand.Read(bytes)
	return "0x" + hex.EncodeToString(bytes)
}

// GetCapitalSupplies returns capital supplies with optional filters
func GetCapitalSupplies(c *gin.Context) {
	userIDStr := c.Query("userId")
	token := c.Query("token")
	status := c.Query("status")

	query := `
		SELECT id, user_id, token, amount, wallet_address, tx_hash, status, created_at, updated_at
		FROM capital_supplies
		WHERE 1=1
	`
	params := []interface{}{}
	paramCount := 1

	if userIDStr != "" {
		userID, err := strconv.Atoi(userIDStr)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid userId"})
			return
		}
		query += fmt.Sprintf(" AND user_id = $%d", paramCount)
		params = append(params, userID)
		paramCount++
	}

	if token != "" {
		query += fmt.Sprintf(" AND token = $%d", paramCount)
		params = append(params, token)
		paramCount++
	}

	if status != "" {
		query += fmt.Sprintf(" AND status = $%d", paramCount)
		params = append(params, status)
		paramCount++
	}

	query += " ORDER BY created_at DESC"

	rows, err := config.DB.Query(query, params...)
	if err != nil {
		log.Printf("Error querying capital supplies: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch capital supplies"})
		return
	}
	defer rows.Close()

	supplies := []map[string]interface{}{}
	for rows.Next() {
		var supply models.CapitalSupply
		err := rows.Scan(
			&supply.ID,
			&supply.UserID,
			&supply.Token,
			&supply.Amount,
			&supply.WalletAddress,
			&supply.TxHash,
			&supply.Status,
			&supply.CreatedAt,
			&supply.UpdatedAt,
		)
		if err != nil {
			log.Printf("Error scanning capital supply: %v", err)
			continue
		}
		supplies = append(supplies, supply.ToResponse())
	}

	c.JSON(http.StatusOK, supplies)
}

// CreateCapitalSupply creates a new capital supply
func CreateCapitalSupply(c *gin.Context) {
	var req CreateCapitalSupplyRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "userId, token, amount, and walletAddress are required",
		})
		return
	}

	if !isValidToken(req.Token) {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": fmt.Sprintf("Invalid token. Must be one of: %v", validTokens),
		})
		return
	}

	var txHash sql.NullString
	if req.TxHash != "" {
		txHash = sql.NullString{String: req.TxHash, Valid: true}
	}

	query := `
		INSERT INTO capital_supplies (user_id, token, amount, wallet_address, tx_hash, status)
		VALUES ($1, $2, $3, $4, $5, 'pending')
		RETURNING id, user_id, token, amount, wallet_address, tx_hash, status, created_at, updated_at
	`

	var supply models.CapitalSupply
	err := config.DB.QueryRow(
		query,
		req.UserID,
		req.Token,
		req.Amount,
		req.WalletAddress,
		txHash,
	).Scan(
		&supply.ID,
		&supply.UserID,
		&supply.Token,
		&supply.Amount,
		&supply.WalletAddress,
		&supply.TxHash,
		&supply.Status,
		&supply.CreatedAt,
		&supply.UpdatedAt,
	)

	if err != nil {
		log.Printf("Error creating capital supply: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create capital supply"})
		return
	}

	log.Printf("Created capital supply %d for user %d: %f %s", supply.ID, req.UserID, req.Amount, req.Token)

	c.JSON(http.StatusCreated, supply.ToResponse())
}

// GenerateDepositAddress generates or retrieves a deposit address
func GenerateDepositAddress(c *gin.Context) {
	var req GenerateDepositAddressRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "userId and token are required",
		})
		return
	}

	if !isValidToken(req.Token) {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": fmt.Sprintf("Invalid token. Must be one of: %v", validTokens),
		})
		return
	}

	// Check for existing active address
	var existing models.DepositAddress
	checkQuery := `
		SELECT id, address, created_at
		FROM deposit_addresses
		WHERE user_id = $1 AND token = $2 AND status = 'active' AND swept = FALSE
		ORDER BY created_at DESC LIMIT 1
	`

	err := config.DB.QueryRow(checkQuery, req.UserID, req.Token).Scan(
		&existing.ID,
		&existing.Address,
		&existing.CreatedAt,
	)

	if err == nil {
		// Return existing address
		c.JSON(http.StatusOK, gin.H{
			"id":        existing.ID,
			"address":   existing.Address,
			"token":     req.Token,
			"createdAt": existing.CreatedAt,
			"isNew":     false,
		})
		return
	}

	if err != sql.ErrNoRows {
		log.Printf("Error checking existing address: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to check existing address"})
		return
	}

	// Generate new address
	depositAddress := generateEthAddress()

	insertQuery := `
		INSERT INTO deposit_addresses (user_id, token, address, status, swept)
		VALUES ($1, $2, $3, 'active', FALSE)
		RETURNING id, user_id, token, address, status, swept, created_at
	`

	var newAddr models.DepositAddress
	err = config.DB.QueryRow(insertQuery, req.UserID, req.Token, depositAddress).Scan(
		&newAddr.ID,
		&newAddr.UserID,
		&newAddr.Token,
		&newAddr.Address,
		&newAddr.Status,
		&newAddr.Swept,
		&newAddr.CreatedAt,
	)

	if err != nil {
		log.Printf("Error creating deposit address: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate deposit address"})
		return
	}

	log.Printf("Generated deposit address %s for user %d: %s", depositAddress, req.UserID, req.Token)

	c.JSON(http.StatusCreated, gin.H{
		"id":        newAddr.ID,
		"userId":    newAddr.UserID,
		"token":     newAddr.Token,
		"address":   newAddr.Address,
		"status":    newAddr.Status,
		"swept":     newAddr.Swept,
		"createdAt": newAddr.CreatedAt,
		"isNew":     true,
	})
}

// GetDepositAddresses returns deposit addresses for a user
func GetDepositAddresses(c *gin.Context) {
	userIDStr := c.Query("userId")
	if userIDStr == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "userId is required"})
		return
	}

	userID, err := strconv.Atoi(userIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid userId"})
		return
	}

	query := `
		SELECT id, user_id, token, address, status, swept, created_at, updated_at
		FROM deposit_addresses
		WHERE user_id = $1
		ORDER BY created_at DESC
	`

	rows, err := config.DB.Query(query, userID)
	if err != nil {
		log.Printf("Error querying deposit addresses: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch deposit addresses"})
		return
	}
	defer rows.Close()

	addresses := []map[string]interface{}{}
	for rows.Next() {
		var addr models.DepositAddress
		err := rows.Scan(
			&addr.ID,
			&addr.UserID,
			&addr.Token,
			&addr.Address,
			&addr.Status,
			&addr.Swept,
			&addr.CreatedAt,
			&addr.UpdatedAt,
		)
		if err != nil {
			log.Printf("Error scanning deposit address: %v", err)
			continue
		}
		addresses = append(addresses, map[string]interface{}{
			"id":        addr.ID,
			"userId":    addr.UserID,
			"token":     addr.Token,
			"address":   addr.Address,
			"status":    addr.Status,
			"swept":     addr.Swept,
			"createdAt": addr.CreatedAt,
			"updatedAt": addr.UpdatedAt,
		})
	}

	c.JSON(http.StatusOK, addresses)
}
